[우아한9조대] 3주차 Facebook 기능 구현

eunsukko opened the pull request:
안녕하세요!
우아한 구조대 3주차 리뷰 요청 드립니다.

- 이미지 저장위치 변경  (로컬 -> s3) 
- 패키지 리팩토링 
- 프로필 페이지 관련 기능
    - 개인정보 수정
    - 프로필 사진 업로드
    - 친구 목록 출력
- 알림 기능 (푸시 사용)
- 비동기 적용
   - 회원가입
    - 로그인 
    - 글, 댓글 출력

감사합니다..!

codeanddonuts committed:
fix/refactor: front-end (#101)

* fix: profile page

* refactor: index page javascript

* fix: attachment modal

* refactor: search

choihz committed:
refactor: 전체적인 리팩토링 (#102)

* refactor: 코드 포맷팅 및 줄바꿈

* refactor: 사용하지 않는 클래스 제거

* refactor: 테스트 리팩토링

vsh123 committed:
refactor: 패키지 구조 리팩토링 (#103)

* refactor: pakage

* remove: 불필요 파일 제거

vsh123 committed:
feat: 테스트 db 연동 (#109)

* refactor: pakage

* remove: 불필요 파일 제거

* feat: 테스트 데이터베이스 설정

choihz committed:
feat: 프로필 수정 버튼 및 폼 (#110)

choihz committed:
feat: 로그아웃 버튼 (#111)

men7627 committed:
feat: 회원가입 및 로그인 비동기 (#114)

* feat: 로그인 및 회원가입 ajax 구현

* feat: 로그인 및 회원가입 ajax 구현

* fix: 오타 수정

* fix: signup form 태그 삭제

choihz committed:
fix: 친구 추가 오류 수정 (#113)

choihz committed:
feat: 프로필 페이지에 친구 목록 출력 (#116)

* feat: 친구 목록 출력

* refactor: 메서드 이름 변경

vsh123 committed:
feat: 프로필 사진/개인정보 수정 (#117)

* feat: 이미지 미리보기 추가

* feat: /api/users/{id}/info 구현

* feat: 사진을 제외한 프로필 수정 기능 구현

* feat: UserProfileImage추가

* fix: 프로필 수정에 따른 image 로딩

* refactor: 업데이트 데이터 확인

* refactor: validataion 추가

* refactor: 붎필요 내용 제거

* fix: 패스워드 확인 로직 수정

codeanddonuts committed:
fix: front-end hotfix (#118)

* fix: front-end

* fix: article editing

vsh123 committed:
feat: 정적파일 저장 S3연동 (#121)

* feat: S3 연동

* refactor: 불필요 메소드 제거

* refactor: attachment service 패키지 생성

* fix: 이미지 저장 경로 수정

codeanddonuts committed:
feat: Push notification (#122)

* feat: notification service & controller

* feat: notification when received friend requests, comments, or likes

* fix: no notifications to own article

choihz committed:
feat: 글 및 댓글 비동기 조회 구현 (#123)

* feat: 뉴스피드 및 프로필 페이지에 비동기로 Article 출력

* refactor: 사용하지 않는 메서드 제거

* refactor: Optimize imports

* feat: 뉴스피드 및 프로필 페이지에 비동기로 Comment 출력

* fix: 이미지 경로 오류 수정

* fix: 테스트 오류 수정

choihz committed:
feat: Article 좋아요 여부 표시 (#125)

codeanddonuts committed:
refactor: Push notification (#126)

* feat: notification service & controller

* feat: notification

* feat: notification when received friend requests, comments, or likes

* fix: no notifications to own article

* fix: broken test

choihz committed:
refactor: 모달 리디자인 (#129)

* refactor: 파비콘 교체

* feat: 프로필 수정 모달 리디자인

* refactor: 사진 첨부 모달 리디자인

vsh123 committed:
feat: 이미지 파일 확장자 검사 기능 구현 (#127)

* setting: apache.tika 의존성 추가

* feat: 이미지 확장자 확인 기능

* fix: 피드백 반영

* fix: multipartfile -> file 변환 시 버그 수정

1. 파일 변환 시 이름 중복을 방지

wmakerjun commented on src/main/resources/static/js/login.js:
```diff
@@ -0,0 +1,29 @@
+        .then(response => {
+            if(response.status === 200){
+                alert('로그인 완료')
+                window.location.href = BASE_URL + '/newsfeed';
```
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun replied:
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/notification.js:
```diff
@@ -0,0 +1,70 @@
+      alert(message.sourceUser.name + " 님이 " + 
+      (() => {
+        switch(message.type) {
+          case "FRIEND_REQUEST":
```
string 문자열을 직접 비교하는 것은 나중에 실수나 오류가 발생할 확률을 높이곤 해요~
따로 상태값을 담는 변수를 만들어서 사용하면 더 좋을 것 같아요 :)

```javascript
const MESSAGE_TYPE = {
	FRIEND_REQUEST: 'FRIEND_REQUEST',
	COMMENT: 'COMMENT'
	...
}
```

그리고 문자열을 합칠 때는  [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

choihz committed:
refactor: 테스트 리팩토링 (#133)

* refactor: 테스트 어노테이션 중복 제거

* refactor: 사용하지 않는 메서드 제거

* refactor: notify() 분리

* test: TotalServiceTest 구현

* test: 친구 불러오기 테스트 구현

* refactor: 테스트 중복 제거

* test: 유저가 작성한 글 조회 테스트 구현

* refactor: 피드백 반영

* refactor: 불필요한 코드 제거

codeanddonuts committed:
test/fix/refactor: Push notification (#130)

* fix: websocket refreshing

* tests: notification

* feat: push notification front-end

* refactor: notification

* feat: separated origin summary and content

codeanddonuts committed:
fix: like push notification (#134)

woowahan-pjs commented on src/main/java/techcourse/fakebook/domain/user/UserProfileImage.java:
```diff
@@ -0,0 +1,33 @@
+public class UserProfileImage {
+    public static String USER_STATIC_FILE_PATH = "user/profile/";
+    public static String DEFAULT_IMAGE_NAME = "default.png";
+    public static String DEFAULT_IMAGE_PATH = "https://woowa-rescue9.s3.ap-northeast-2.amazonaws.com/user/profile/default.png";
```
s3 도메인은 프로퍼티로 관리할 수 없을까요?

phs1116 replied:
찬성합니다. 또한 해당 값들이 현재 이 클래스가 직접 알 필요가 없어보이네요.

vsh123 replied:
해당 내용과 관련하여 질문이 있습니다!
우선 알려주신 피드백을 반영하기 위해 s3.properties라는 프로퍼티를 생성하여 위의 상수들을 모두 프로퍼티 안으로 옮겼습니다.
이후 해당 값들을 @Value 어노테이션을 사용하여 받는 과정에서 두가지 이슈 및 고민거리가 생겼는데 해당 부분에 대해서의견 주시면 감사하겠습니다.

1. properties말고 yml로는 방법이 없는건지..?(여러 방법을 사용해봐도 spring boot에서 load를 하지 못하더라구요 ㅠㅠ)

2. 현재 상수들을 다음 이미지와 같이 S3Uploader의 필드로 가지고 있습니다.
![image](https://user-images.githubusercontent.com/32254689/64076185-78aded00-ccfc-11e9-85ee-d94a90ec1f82.png)
하지만 해당 상수들은 아래와 같이 attachmentService에서 사용하고 있습니다.
![image](https://user-images.githubusercontent.com/32254689/64076219-f70a8f00-ccfc-11e9-9003-7bf4ac775053.png)

그래서 해당 상수들을 어디에 선언해주는 것을 고민해본 결과 다음과 같이 4가지의 위치를 생각해 보았는데 이에 대해서 더 좋은 의견 있으시면 주시면 감사하겠습니다!
(1) 지금과 같이 S3uploader에 선언
(2) AttachmentService에 선언
(3) AwsConfig파일에 각각의 String을 Bean으로 선언하여 관리
(4) AwsConfig파일에 각각의 String을 static으로 선언하여 관리

감사합니다!

vsh123 replied:
해당 내용과 관련하여 질문이 있습니다!
우선 알려주신 피드백을 반영하기 위해 s3.properties라는 프로퍼티를 생성하여 위의 상수들을 모두 프로퍼티 안으로 옮겼습니다.
이후 해당 값들을 @Value 어노테이션을 사용하여 받는 과정에서 두가지 이슈 및 고민거리가 생겼는데 해당 부분에 대해서의견 주시면 감사하겠습니다.

1. properties말고 yml로는 방법이 없는건지..?(여러 방법을 사용해봐도 spring boot에서 load를 하지 못하더라구요 ㅠㅠ)

2. 현재 상수들을 다음 이미지와 같이 S3Uploader의 필드로 가지고 있습니다.
![image](https://user-images.githubusercontent.com/32254689/64076185-78aded00-ccfc-11e9-85ee-d94a90ec1f82.png)
하지만 해당 상수들은 아래와 같이 attachmentService에서 사용하고 있습니다.
![image](https://user-images.githubusercontent.com/32254689/64076219-f70a8f00-ccfc-11e9-9003-7bf4ac775053.png)

그래서 해당 상수들을 어디에 선언해주는 것을 고민해본 결과 다음과 같이 5가지의 위치를 생각해 보았는데 이에 대해서 더 좋은 의견 있으시면 주시면 감사하겠습니다!
(1) 지금과 같이 S3uploader에 선언
(2) AttachmentService에 선언
(3) AwsConfig파일에 각각의 String을 Bean으로 선언하여 관리
(4) AwsConfig파일에 각각의 String을 static으로 선언하여 관리
(5) Uploader라는 S3Uploader 인터페이스에 선언

감사합니다!

vsh123 replied:
해당 내용과 관련하여 질문이 있습니다!
우선 알려주신 피드백을 반영하기 위해 upload.properties라는 프로퍼티를 생성하여 위의 상수들을 모두 프로퍼티 안으로 옮겼습니다.
이후 해당 값들을 @Value 어노테이션을 사용하여 받는 과정에서 두가지 이슈 및 고민거리가 생겼는데 해당 부분에 대해서의견 주시면 감사하겠습니다.

1. properties말고 yml로는 방법이 없는건지..?(여러 방법을 사용해봐도 spring boot에서 load를 하지 못하더라구요 ㅠㅠ)

2. 현재 상수들을 다음 이미지와 같이 UploaderConfig의 필드로 가지고 있습니다.
![image](https://user-images.githubusercontent.com/32254689/64076565-5ec2d900-cd01-11e9-9779-152f7f601fb2.png)
그리고 attachmentService에서 UploaderConfig를 의존하여 사용하고 있습니다.

이렇게 하고보니 attachmentService에 UploaderConfig라는 의존이 하나 더 추가 된 것 같아 다음과 같이 위치에 대한 고민을 해보았는데 의견주시면 감사하겠습니다!
(1) 지금과 같이 UploaderConfig에 선언
(2) AttachmentService에 선언
(3) S3Uploader에 선언 -> 이렇게 하면 Uploader라는 인터페이스를 선언한 의미가 없지 않나 생각이 들기도 합니다!

감사합니다!

woowahan-pjs replied:
아래의 글을 참고하세요.
https://www.baeldung.com/configuration-properties-in-spring-boot

woowahan-pjs commented on src/main/java/techcourse/fakebook/service/article/ArticleService.java:
```diff
@@ -23,18 +31,24 @@
 @Service
 @Transactional
 public class ArticleService {
```
`ArticleService`를 관심사에 맞게 분리할 때가 되었네요.

woowahan-pjs commented on src/main/java/techcourse/fakebook/service/article/ArticleService.java:
```diff
@@ -96,7 +113,14 @@ public boolean like(Long id, UserOutline userOutline) {
             articleLikeRepository.delete(articleLike.get());
```
`unlike()`, `cancelLike()` 등으로 분리!

woowahan-pjs commented on src/main/java/techcourse/fakebook/service/article/ArticleService.java:
```diff
@@ -96,7 +113,14 @@ public boolean like(Long id, UserOutline userOutline) {
-        articleLikeRepository.save(new ArticleLike(userService.getUser(userOutline.getId()), getArticle(id)));
+        final Article article = getArticle(id);
+        articleLikeRepository.save(new ArticleLike(userService.getUser(userOutline.getId()), article));
+        if (article.isNotAuthor(userOutline.getId())) {
```
본인이 '좋아요'를 한 경우에는 알림이 안 가는군요! 👍

woowahan-pjs commented on src/main/java/techcourse/fakebook/service/article/TotalService.java:
```diff
@@ -16,24 +19,15 @@
     private final ArticleAssembler articleAssembler;
 
-    public TotalService(ArticleService articleService, CommentService commentService, UserService userService, FriendshipService friendshipService, ArticleAssembler articleAssembler) {
+    public TotalService(ArticleService articleService, CommentService commentService, UserService userService,
```
`TotalService`? 클래스 이름에 대해 조금 더 고민해 봐요!

woowahan-pjs commented on src/main/java/techcourse/fakebook/utils/s3/S3Uploader.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Optional;
+
+@Component
+public class S3Uploader {
```
`S3Uploader`에서 의미 있는 인터페이스를 도출해 보세요. 어떤 장점이 있을까요?

vsh123 replied:
upload를 인터페이스에서 정의를 해주게 된다면 s3가 아닌 다른 정적 파일 서버로 변화를 하였을 때도 attachmentSerivce에 생성자 주입만 변경을 해주면 되는 이점이 생길 것 같아요! 혹시 이외에도 좋은 장점이 있을까요?

woowahan-pjs reviewed (changes requested):
Fakebook도 많은 기능이 추가되면서 Legacy의 길을 가고 있네요. 👍 
지극히 정상이고, 서비스가 잘 되고 있다는 뜻이기도 하니깐 큰 걱정은 하지 마세요.
크게 두 가지를 조심하면 되겠어요.

1. 컴포넌트를 생성하기 위해 필요한 컴포넌트의 수가 많아진다면 새로운 컴포넌트를 만들 때가 온 것이다.
2. 메서드는 한 가지 일만 해야 한다.

codeanddonuts committed:
feat: Interval between push notifications (#136)

* fix: like push notification

* feat: interval between notifications

choihz committed:
feat: 댓글 좋아요 구현 (#135)

* fix: 테스트 오류 수정

* refactor: 불필요한 로직(recentDate) 제거

* feat: 댓글 좋아요 구현

eunsukko committed:
feat: 친구추천 기능 (#132)

* feat: 친구 추천 서비스 구현

* feat: 친구 추천 프론트 구현

* fix: FriendRecommendationService 로직 버그 수정

* fix: logger 수정

* fix: 중복되는 친구추천 수정

phs1116 commented on src/main/java/techcourse/fakebook/service/article/ArticleService.java:
```diff
@@ -54,26 +65,22 @@ public ArticleResponse findById(Long id) {
         return articleAssembler.toResponse(article, attachments);
     }
 
+    @Transactional
```
클래스에 이미 Transactional을 작성하셔서 생략해도 무방해보여요 :)

phs1116 commented on src/main/java/techcourse/fakebook/service/attachment/AttachmentService.java:
```diff
@@ -0,0 +1,101 @@
+
+            return getAttachmentResponse(articleAttachmentRepository.save(articleAttachment));
+        } catch (IOException e) {
+            throw new FileSaveException();
```
어차피 S3Uploader에서의 예외가 복구 불가능한 예외면 서비스가 아닌 Uploader에서 포장해주는게 좋을 것 같아요.
S3Uploader를 사용하는 곳마다 매번 예외 전환을 하는것 보다는요. :)
그리고 기존 예외에 대한 사유를 정확하게 알기위해 예외 메시지를 전달하거나, 예외를 포장하거나, 에러 로그를 찍는게 좋지 않을까요?
현재는 FileSaveException가 던져진다는것 외에는 정확한 사유를 알 수 없어보여요.

vsh123 replied:
S3Uploader에서 예외를 포장하고 실패가 난 부분에서 error 로그를 찍는 방법으로 변경하였습니다!

phs1116 commented on src/main/java/techcourse/fakebook/service/attachment/AttachmentService.java:
```diff
@@ -0,0 +1,101 @@
+
+        String mimeType = tika.detect(file.getBytes());
+        log.debug("### MIME Type = {}", mimeType);
+        if (!mimeType.startsWith("image")) {
```
상수처리 :)

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -1,7 +1,8 @@
+"use strict";
 
+const App = (() => {
   const BASE_URL = "http://" + window.location.host
```
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -56,7 +57,96 @@ const App = (() => {
+
   class ArticleService extends Service {
+    async showNewsfeed() {
+      this.show(BASE_URL + "/api/articles")
```
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -56,7 +57,96 @@ const App = (() => {
+    }
+
+    async showArticles(userId) {
+      this.show(BASE_URL + "/api/users/" + userId + "/articles")
```
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -56,7 +57,96 @@ const App = (() => {
+    }
+
+    async checkLike(articleId) {
+      const isLiked = (await axios.get(BASE_URL + "/api/articles/" + articleId + "/like")).status
```
url은 [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/app.js:
```diff
@@ -1,7 +1,8 @@
-  "use strict"
+"use strict";
 
+const App = (() => {
```
1. app.js 파일에 많은 클래스가 선언되어있는 것 같아요. 객체별로 파일을 분리해보면 어떨까요?
2. 문자열을 결합하는 방식이 "" + "" 이렇게 많이 하고 있는데요. [문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/notification.js:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    dispatch(message) {
+      const srcUserName = message.srcUser.name + " 님"
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/notification.js:
```diff
@@ -0,0 +1,106 @@
+    dispatch(message) {
+      const srcUserName = message.srcUser.name + " 님"
+      switch(message.type) {
+        case "CHAT":
```
string 문자열을 직접 비교하는 것은 나중에 실수나 오류가 발생할 확률을 높이곤 해요~
따로 상태값을 담는 변수를 만들어서 사용하면 더 좋을 것 같아요 :)

```javascript
const MESSAGE_TYPE = {
	CHAT: 'CHAT',
	FRIEND_REQUEST: 'FRIEND_REQUEST'
	...
}
```

wmakerjun commented on src/main/resources/static/js/notification.js:
```diff
@@ -0,0 +1,106 @@
+      const srcUserName = message.srcUser.name + " 님"
+      switch(message.type) {
+        case "CHAT":
+          this.pushToQueue(message.srcUser.name + " : " + message.content)
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/notification.js:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    async requestNewChannelAddress() {
+      return (await axios.get("http://" + window.location.host + MESSAGE_BROKER_URI)).data.address
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/static/js/signup.js:
```diff
@@ -0,0 +1,53 @@
+        const email = document.getElementById('new-email').value;
+        const password = document.getElementById('new-password').value;
+        const gender = document.getElementById('female').checked ? 'female' : 'male';
+        const birth = document.getElementById("birth-year").value + "/" +
```
[문자열 리터럴](https://poiemaweb.com/es6-template-literals)을 참고해서 표현하면 좋을 것 같아요!

wmakerjun commented on src/main/resources/templates/candidates.html:
```diff
@@ -0,0 +1,145 @@
+        <div class="mrg-top-20"></div>
+    </footer>
+</div>
+<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
```
html 페이지에서 불러오는 script tag들이 중복이 있는데요~ 타임리프를 이용해서 중복을 제거하고 재사용하면 어떨까요~?

wmakerjun commented on src/main/resources/templates/candidates.html:
```diff
@@ -0,0 +1,145 @@
+<!DOCTYPE html>
+<html lang="ko">
+<head>
```
head 부분에 import 하는 부분들이 중복이 많은데 타임리프를 이용해서 재사용하는 방식으로 리팩토링 해보면 어떨까요~?

wmakerjun reviewed (changes requested):
app.js에 class 기반으로 짠 코드를 보고 깜짝 놀랬습니다!! 💯 
2~3가지만 수정하면 더 멋진 코드가 될 것 같아요 :) 얼마 남지 않은 시간도 화이팅입니다 😄 

1. 문자열 결합시 [문자열 리터럴](https://poiemaweb.com/es6-template-literals) 활용
2. html 페이지에서 script와 head 중복 제거
3. 가능하면 app.js 파일을 분리

phs1116 commented on src/main/java/techcourse/fakebook/service/attachment/AttachmentService.java:
```diff
@@ -0,0 +1,101 @@
+
+        List<String> splits = Arrays.asList(originalFileName.split("\\."));
+        if (splits.size() < 1) {
+            throw new FileSaveException();
```
S3에서 업로드를 실패한 경우, 혹은 파일 이름 문제등에 대해서 실패 사유를 추적하기 상당히 어려워보여요.
같은 예외를 던지더라도 메시지 및 로그로 표현해야 이슈 시 파악이 쉬울듯 보입니다.

phs1116 commented on src/main/java/techcourse/fakebook/service/comment/CommentService.java:
```diff
@@ -27,14 +30,17 @@
     private final CommentRepository commentRepository;
     private final CommentLikeRepository commentLikeRepository;
     private final CommentAssembler commentAssembler;
+    private final NotificationService notificationService;
```
이 클래스도 이제 의존이 상당히 많네요.
역할에 맞게 분리해보는게 좋을 것 같아요.

phs1116 commented on src/main/java/techcourse/fakebook/service/friendship/FriendCandidateFactory.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    // TODO: 이름을 좀 더 명확하게
+    public List<FriendCandidate> createCandidates(Long userId) {
```
친구와 친구의 친구들, 방문자수가 많으면 많을수록 너무 한꺼번에 많은 데이터를 처리할 것 같은데요.
성능에 영향을 덜 줄수 있는 다른 방법도 고민해보면 어떨까요?
예를들어 전체를 다 계산하지 않고 후보에 대해서 일부만 제공한다던가요.

phs1116 commented on src/main/java/techcourse/fakebook/service/friendship/VisitorTrackingService.java:
```diff
@@ -0,0 +1,37 @@
+public class VisitorTrackingService {
+    private static final Logger log = LoggerFactory.getLogger(VisitorTrackingService.class);
+
+    private final Map<Long, Counter<Long>> counters = new HashMap<>();
```
방문자 카운터가 배포시마다 초기화가 되거나, 서버마다 다른 데이터를 보여 줄 수 있겠네요. :)
지금에 고칠건 아니지만 그냥 한번 이런 경우 어떻게 하면 좋았을지에 대해 고민해보는것도 좋을 것 같습니다.

phs1116 commented on src/main/java/techcourse/fakebook/service/friendship/MutualFriendRankingStrategy.java:
```diff
@@ -0,0 +1,17 @@
+import java.util.stream.Collectors;
+
+@Component
+public class MutualFriendRankingStrategy {
```
MutualFriendWithVisitorRankingStrategy, MutualFriendRankingStrategy 둘 다 같은 역할의 전략으로 보이는데 공통 인터페이스로 추출해보면 어떨까요?

phs1116 commented on src/main/java/techcourse/fakebook/service/friendship/FriendRecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+
+        return rankedFriendCandidates.stream()
+                .filter(candidate -> userOutlines.containsKey(candidate.getFriendId()))
+                .map(candidate -> new FriendRecommendation(userOutlines.get(candidate.getFriendId()), candidate.getMutualFriendIds()))
```
MutualFriend가 맞나요?
그저 방문자에 대한 정보도 포함되있어 보이는데요. :)

phs1116 commented on src/main/java/techcourse/fakebook/service/friendship/FriendRecommendationService.java:
```diff
@@ -0,0 +1,63 @@
+                .collect(Collectors.toList());
+
+        Map<Long, UserOutline> userOutlines = new HashMap<>();
+        for (UserOutline outline : userService.findFriends(candidateIds)) {
```
friendCandidateFactory.createCandidates에서 이미 한번씩 대량으로 친구 정보를 가져오는데
여기에서도 다시 한번 대량으로 조회하네요.
이 점도 한번 고민해보면 좋을 것 같아요.

phs1116 commented on src/main/java/techcourse/fakebook/service/notification/NotificationChannelMapper.java:
```diff
@@ -0,0 +1,26 @@
+public class NotificationChannelMapper {
+    private final Map<Long, NotificationChannel> channels = new ConcurrentHashMap<>();
+
+    public NotificationChannel assignTo(long userId) {
```
여기서도 지금 당장의 솔루션으로는 해결할 수 없지만 생각해볼 거리 하나.
서버가 여러대일 경우, 서버가 순차적으로 배포가 될 경우 이슈가 발생 할 수 있어요.

phs1116 commented on src/main/java/techcourse/fakebook/service/notification/assembler/NotificationAssembler.java:
```diff
@@ -0,0 +1,61 @@
+
+@Component
+public class NotificationAssembler {
+    private int maxSummaryLength = 15;
```
컴포넌트인데 maxSummaryLength가 상수가 아닌 이유가 있을까요?
또한 이 값에 대해서 변경가능하도록 열어놓은 이유도 알 수 있을까요?

phs1116 commented on src/main/java/techcourse/fakebook/service/notification/NotificationService.java:
```diff
@@ -0,0 +1,58 @@
+import techcourse.fakebook.service.notification.dto.NotificationResponse;
+
+@Service
+public class NotificationService {
```
다른 서비스에서 NotificationService를 사용 시 NotificationService에서 에러가 났을 경우
NotificationService를 사용한 서비스가 트랜잭션 내에 속해있다면 롤백이 될 위험이 있습니다.
또한 NotificationService에 지연이 있을 경우 사용자는 타임아웃을 받을 수도 있어요.

사실 알림은 가지 않더라도 크게 이슈가 없는 기능이다 보니까 기존에 동작하는 트랜잭션이랑은 별도로 가는게 맞구요.


따라서 해당 서비스에 대해서는 별도 트랜잭션으로 관리하거나 `@Async`를 통해서 비동기로 관리해보면 어떨까요?

phs1116 commented on src/main/java/techcourse/fakebook/web/controller/article/ArticleApiController.java:
```diff
@@ -16,14 +18,16 @@
-    public ResponseEntity<List<ArticleResponse>> findAll() {
-        List<ArticleResponse> articles = articleService.findAll();
+    public ResponseEntity<List<TotalArticleResponse>> findAll(@SessionUser UserOutline userOutline) {
+        List<TotalArticleResponse> articles = totalService.findArticlesByUserIncludingFriendsArticles(userOutline.getId());
```
페이지네이션에 대해서도 고민해보면 어떨까요?

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
한주 차이동안 작성한 코드량이 상당히 많네요. 👍 
저번 단계에서 말했다시피 하나의 서비스가 처리하는 책임과 역할이 점점 많아지고 있어 보입니다.
이에 대해서 고민해보면 좋을 것 같구요.
예외처리시 에러로그도 잘 신경써주시면 좋을 것 같습니다.

그 외에 운영되는 서비스가 아니다보니 상관은 없지만 
문제가 있을만하고, 고민해보면 좋을 부분에 대해서 대해서 몇가지 피드백 추가드렸어요.
참고 부탁드릴게요. :)

men7627 committed:
feat: 채팅 기능 구현 및 읽음 구현 (#147)

* 채팅 구현하는 중

* 임시 저장

* fix: 사용하지 않는 테스트 제거

* fix: 채팅 읽음표시 및 채팅 낚아채기 버그 수정

* fix: 채팅 기능 인터셉터등록

vsh123 committed:
fix: 버그 수정 (#151)

* refactor: 이미지 업로드 실패시 alert 한번만 나오게 수정

* refactor: 프로필 수정시에 빈칸빼고 수정하게 변경

* refactor: 이미지만 있어도 게시글 업로드 가능하게 설정

* fix: 컨벤션 및 오타 수정

vsh123 committed:
refactor: 3주차 피드백 반영(쉐이크반) (#157)

* refactor: 공통 exception 처리에 따른 error logging

* refactor: s3업로드 관련 상수 프로퍼티로 선언

* refactor: Uploader 인터페이스 선언 및 상수 프로퍼티 관리

* refactor: uploader 생성자 주입 변경

* refactor: upload.properties 수정

men7627 committed:
refactor: 채팅, 읽음 refactor & testing (#160)

1. user/userId 인터셉터 처리

vsh123 commented:
Article, Comment 각가의 Service에서 계층을 두어서 리팩토링을 진행하였습니다. 
리팩토링 진행 후 의존이 분리되는 느낌은 있지만 뭔가 리뷰어분들께서 원하시는 방식은 아닌 것 같아 이렇게 글을 남깁니다!
보신 후 미흡한 사항이 보인다면 많은 피드백 부탁드리겠습니다 :) 마지막까지 잘 부탁드리겠습니다!

choihz committed:
refactor: 메신저 프론트 리팩토링 (#161)

* refactor: 피드백 반영

* refactor: 메신저 프론트 리팩토링

vsh123 committed:
refactor: Article,Comment Service 리팩토링 (#162)

* refactor: cancelLike 분리

* refactor: articleLikeService 생성

* refactor: ArticleInnerService 생성

* refactor: CommentLikeService

* refactor: CommentInnerService

* refactor: 컨벤션

men7627 committed:
refactor: test 오류 수정 (#165)

codeanddonuts committed:
refactor: async push notification (#164)

* fix: like push notification

* test: method name

* refactor: async push notifications

choihz committed:
feat: 댓글 삭제 구현 (#167)

* feat: 댓글 삭제 구현

* feat: 글 및 댓글 수정/삭제 버튼 노출 변경

vsh123 committed:
fix: timezone (#168)

eunsukko committed:
fix: 로그설정 수정 및 친구추가 링크 위치 변경 (#169)

* fix: logback.xml 로그 설정 수정

문제상황
- 현재 root 와 techcourse 로그가 중복되어서 찍히고 있음
- root 의 경우 techcourse 내부의 로그들을 출력하지 못하고 있음

해결방식
- techcourse 내부에서 additivity="false" 적용(부모에서 로그를 남기지 않도록)

* fix: 친구추가 링크 위치 변경

phs1116 commented on src/main/java/techcourse/fakebook/service/article/ArticleService.java:
```diff
@@ -0,0 +1,83 @@
+public class ArticleService {
+    private static final Logger log = LoggerFactory.getLogger(ArticleService.class);
+
+    private final ArticleInnerService articleInnerService;
```
ArticleInnerService가 필요한가요?
그냥 바이패스하는 서비스로 보이는데
ArticleInnerService를 ArticleService로,
사용하는 측에서 ArticleService와 ArticleLikeService를 사용하면 안되나요?
의존은 Article에 대한 기능들은 느슨하게 가져가지만 ArticleService의 관심사가 많은건 변함 없어보이네요. :)

phs1116 commented on src/main/java/techcourse/fakebook/service/attachment/AttachmentService.java:
```diff
@@ -0,0 +1,103 @@
+    private final ArticleAttachmentRepository articleAttachmentRepository;
+    private final AttachmentAssembler attachmentAssembler;
+    private final Uploader uploader;
+    private final UploaderConfig uploaderConfig;
```
UploaderConfig라는것 자체가 Uploader에 종속적인 내용인데 여기서 주입받을 필요가 있을까요?
Uploader의 구현체에서 주입받으면 되지 않을까 싶습니다. :)

phs1116 commented on src/main/java/techcourse/fakebook/service/attachment/AttachmentService.java:
```diff
@@ -0,0 +1,103 @@
+    private final UploaderConfig uploaderConfig;
+
+    public AttachmentService(ArticleAttachmentRepository articleAttachmentRepository, AttachmentAssembler attachmentAssembler,
+                             S3Uploader uploader, UploaderConfig uploaderConfig) {
```
생성자에서 주입받을때는 구현체가 아닌 인터페이스 기반으로 주입받도록 해주세요. :)
그래야 DI의 장점을 살릴 수 있습니다. :) 왜 그런지에 대해서도 고민해보시면 좋을 것 같아요.

phs1116 commented on src/main/java/techcourse/fakebook/service/comment/CommentService.java:
```diff
@@ -0,0 +1,87 @@
+public class CommentService {
+    private static final Logger log = LoggerFactory.getLogger(CommentService.class);
+
+    private final CommentInnerService commentInnerService;
```
ArticleService와 마찬가지로 CommentLikeService와 CommentService를 별도로 사용하면 되지 않을까요?
단일 책임에 대해서 고민해보면 좋을 것 같아요.

phs1116 commented on src/main/java/techcourse/fakebook/utils/uploader/s3/S3Uploader.java:
```diff
@@ -0,0 +1,75 @@
+        return amazonS3Client.getUrl(bucket, fileName).toString();
+    }
+
+    private void removeNewFile(File targetFile) {
```
targetFile.delete() 가 false인 상황을 예외로 던지거나 에러로그를 찍고 있지 않은것 같은데요.
그럼 이슈사항이 아닌 케이스 같은데 이렇게 로그를 찍는 메서드가 의미가 있나요?

phs1116 commented on src/main/java/techcourse/fakebook/service/article/TotalService.java:
```diff
@@ -16,24 +19,15 @@
-                .collect(Collectors.toList());
-    }
-
     public List<TotalArticleResponse> findArticlesByUser(Long userId) {
```
TotalService보다는 해당 데이터의 목적성에 맞는 이름으로 서비스를 분리해도 괜찮을 것 같아요.
findArticlesByUserIncludingFriendsArticles같은 경우 현재는 NewsFeed를 위한 데이터로 보이니
NewsFeedService가 될수 있을거구요
findArticlesByUser는 UserArticleFacade라고 할수도 있겠네요.

phs1116 commented on src/main/java/techcourse/fakebook/utils/uploader/s3/S3Uploader.java:
```diff
@@ -0,0 +1,75 @@
+
+            return upload(uploadFile, dirName, fileName);
+        } catch (IOException e) {
+            log.error("FileSaveError : file write 실패");
```
e.getMessage등도 로그에 찍는게 좋을 것 같네요.:)

phs1116 commented on src/main/java/techcourse/fakebook/service/attachment/AttachmentService.java:
```diff
@@ -0,0 +1,103 @@
+                throw new NotImageTypeException();
+            }
+        } catch (IOException e) {
+            log.error("FileSaveError : file.getByte 실패");
```
정확히 어떤 사유로 실패를 했는지 파악하기 위해서 e.getMessage도 찍어보아요 :)

phs1116 reviewed (changes requested):
안녕하세요. 리뷰어 화투입니다.
꼭 서비스의 이름이 엔티티랑만 연관될 필요는 없어요.
이 객체가 책임지고 있는 비즈니스에 맞는 네임으로 짓고 
이름에 맞는 책임만 짓도록 하는것도 좋아요.

그 외에 몇가지 피드백도 추가했는데 참고 부탁드릴게요. :)

vsh123 committed:
Revert "refactor: async push notification (#164) ~ (#169)"

This reverts commit bd94cca7e1e98749267d331112537390f9e4840f.

choihz committed:
fix: 최종 수정 (#171)

* feat: 댓글 삭제 구현

* feat: 글 및 댓글 수정/삭제 버튼 노출 변경

* fix: timezone

* fix: 친구 추가 링크 위치 변경

* fix: logback.xml 로그 설정 수정

* fix: 변수명 변경

* fix: 인터셉터 수정

* fix: 오류 수정

woowahan-pjs reviewed (approved):
축하합니다! 백엔드 Java 레벨 2의 모든 미션 완료하였어요. 🥇 
나만의 블로그부터 미니 프로젝트까지
수많은 미션을 수행하느라 대단히 고생 많으셨습니다. 💯 
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.
이만 Merge 하겠습니다.
